rfcs/90-parse-pairs-from-string.md

```diff
@@ -0,0 +1,106 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-06-01
+deprecates:
+  - 13-remove-config-string.md
+---
+
+# GSP-90: Parse Pairs from String
+
+## Background
+
+We had config strings like `s3://bucket_name?access_key=xxxx&secret_key=xxxx` before. This is introduced in [GSP-3] but deprecated by [GSP-13].
+
+There are two reasons why config strings are removed, as mentioned in [GSP-13]:
+
+1. Usability: If an application uses [go-storage]'s config string, then it either exposes the format of config string to end user directly (additional burden for end users) or it has to manipulate its config into our format (additional burden for application developers). And if the application wants to combine its other config into a config string, this is also inconvenient.
+2. Implementation: We didn't find a good way to parse pairs from strings easily and safely.
+
+So we let the users of [go-storage] construct pairs directly using `WithXxx()`. 
+
+But passing string config is indeed more convenient, if not much. And actually parsing string into pairs is not that difficult. We can use a registry mechanism like [GSP-48]'s implementation.
+
+## Proposal
+
+So I propose the following mechanism for parsing pairs:
+
+### Pair Registry
+
+We first register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+type PairMap map[string]reflect.Type
```

Comment:
Why we need to use `reflect.Type` here? The simple string maybe look enough to us? We only need to support limited types as described in specs.

Reply:
My fault. I didn't recall there's a mapping from string to type specified in specs